js/libs/core.js

```diff
@@ -75,6 +75,76 @@ Fliplet.FormBuilder = (function() {
         };
       }
 
+      component.methods.setSourceValue = function(data) {
+        var result;
+        var $vm = this;
+
+        switch (data.source) {
```

Comment:
Has this been copied from somewhere else? I'm asking because if that's the case we may be able to put this logic into a shared file, e.g. `fliplet-core`

Reply:
@squallstar Yes, it was copied from this [function ](https://github.com/Fliplet/fliplet-widget-form-builder/blob/master/js/libs/form.js#L840). But we changed it a little for our case (striped case function and literal from it.)

Reply:
Can the two functions be aggregated into a single one which we can then move as required later? We should be always working towards keeping code DRY.

Reply:
@squallstar Yes, it is possible. Should I change it so it would work in both cases?

Reply:
Yes please.

Reply:
@squallstar While I've tried to aggregated functions into one I've stuck with a question.
To get data from the Profile in the `.set()` function we use this [code](https://github.com/Fliplet/fliplet-widget-form-builder/blob/master/js/libs/form.js#L861) 
But in our new feature, we will use this [code](https://github.com/Fliplet/fliplet-widget-form-builder/pull/355/files#diff-962c976e04650f0e20b7b5a03282b499b82960438ed596001139d959084b8e2fR88)
In aggregated function what should be left?
And if we take a new example won't it make trouble for the current users?

Reply:
@YaroslavOvdii I think you should use the newest function's logic but with the following change: `Fliplet.Profile.get(data.key);` should always be used as a fallback when the data is not found in the target passport, e.g.

```
value = value from dataSource or saml2 or flipletLogin session
if value is undefined, value = Fliplet.Profile.get(data.key);
return value
```

Does it make sense?

@tonytlwu do you agree with this?

Reply:
@squallstar What you've outlined here makes sense. The LFD was updated with a similar logic.

See https://github.com/Fliplet/fliplet-widget-dynamic-lists/pull/464/files#diff-1e20a0ee0fa5c5689cd8541a5892e5e23248b3bf2ab918043e525c443856f18c

Reply:
@squallstar I'm not sure where should I put the aggregated function so I put it in this comment.
```
/**
 * This function is responsible for getting the value from the source that was specified by the user
 * 
 * @param {Object} data - data that was passed by a user into the set function
 * 
 * @return {Promise} - returns promise with data that was returned from the source
 */
function getSourceData(data) {
  var result;
  var data = options.data;

  if (typeof data === 'function') {
    data = { source: 'function', key: data };
  }

  if (typeof data !== 'object') {
    data = { source: 'literal', key: data };
  }

  switch (data.source) {
    case 'profile':
      if (!data.key) {
        throw new Error('A key is required to fetch data from the user\'s profile');
      }

      result = Fliplet.User.getCachedSession()
        .then(function(session) {
          if (session && session.entries) {
            if (session.entries.dataSource) {
              return session.entries.dataSource.data[data.key];
            }

            if (session.entries.saml2) {
              return session.entries.saml2.data[data.key];
            }

            if (session.entries.flipletLogin) {
              return session.entries.flipletLogin.data[data.key];
            }
          }

          return Fliplet.Profile.get(data.key);
        });
      
        break;
    case 'query':
      if (!data.key) {
        throw new Error('A key is required to fetch data from the user\'s profile');
      }
      
      result = Fliplet.Navigate.query[data.key];
      
      break;
    case 'storage':
    case 'appstorage':
      if (!data.key) {
        throw new Error('A key is required to fetch data from the user\'s profile');
      }

      var storage = data.source === 'storage'
        ? Fliplet.Storage
        : Fliplet.App.Storage;
      result = storage.get(data.key);
      
      break;
    case 'function':
      result = data.key();
      
      break;
    case 'literal':
      result = data.key;
      
      break;
    default:
      result = data;

      break;
  }

  if (!(result instanceof Promise)) {
    result = Promise.resolve(result);
  }

  return result;
}
```
How it should work:
When I've analyzed the `set` and my functions I understand that the above part is the same for both of them so I've separated it and when we put it into the `fliplet-core` we should be able to work with the returned value as we had previously but without duplicating the code.